### ISSUE TITLE
Feature Request: Support embeddings as a native input modality in the Messages API

### ISSUE THREAD

**Issue by jmm2020** [anthropics/anthropic-sdk-python#1351]
### The Problem — The Lossy Round-Trip

Every retrieval-augmented system today follows the same pipeline:

```
text → embedding → vector search → retrieve text → feed to LLM
```

The embedding step captures semantic meaning in high-dimensional space (e.g., 4096 dimensions). The retrieval step then **converts it back to flat text** — discarding the geometric relationships, cluster positions, and distance signals that the vector space already computed. The LLM then re-encodes that text into its own internal representations, reconstructing what the embedding already knew.

This is a lossy round-trip. The information exists in vector form, gets serialized to text, and then gets re-vectorized internally by the model. The intermediate text step is a bottleneck — both in fidelity and in token cost.

### The Proposal — Vector Prompt Interface

**What if the Messages API accepted embedding vectors as a native input modality, alongside text and images?**

Conceptually, this would look like a new content block type:

```python
response = client.messages.create(
    model="claude-opus-4-6",
    messages=[{
        "role": "user",
        "content": [
            # Traditional text context
            {"type": "text", "text": "Given these retrieved memories, answer my question:"},
            # NEW: embedding vectors injected at the input-encoding layer
            {
                "type": "embedding",
                "vectors": [
                    {"data": [0.0234, -0.0891, ...],  "dimensions": 4096, "label": "memory_1"},
                    {"data": [0.0112, -0.0453, ...],  "dimensions": 4096, "label": "memory_2"},
                ],
                "model": "qwen3-embedding-8b",  # or voyage-3, text-embedding-3-large, etc.
                "metadata": {
                    "distances": [0.12, 0.34],  # cosine distances from query
                    "retrieval_scores": [0.95, 0.82]  # fusion scores if available
                }
            },
            {"type": "text", "text": "What was the key breakthrough last week?"}
        ]
    }]
)
```

### Why This Matters

1. **Lossless retrieval context** — Geometric relationships between retrieved items (cluster distances, traversal paths, similarity scores) arrive intact instead of being serialized to text descriptions.

2. **Token efficiency** — A 4096-dimensional embedding carries the semantic weight of thousands of tokens in a single vector. Systems with 32K+ retrievable items could provide richer context without hitting token limits.

3. **Native agent-to-agent communication** — Multi-agent systems increasingly use embeddings as inter-agent signals (e.g., binary-quantized vectors streamed via Kafka). Accepting these natively eliminates the serialization/deserialization overhead.

4. **Retrieval metadata preservation** — Fusion scores, distance metrics, graph traversal paths, and other retrieval signals could be passed directly rather than described in natural language.

### Precedent

Images proved that LLMs can process non-text modalities at the input-encoding layer. The architecture already supports multimodal input. Embeddings are semantically closer to the model's internal representations than pixels are — they're a more natural fit for this pattern.

### Real-World Use Case

We operate a multi-agent system (UCIS) with:
- **32,000+ memories** in a graph database, each with 4096d embeddings
- **6-signal retrieval fusion** (vector search, keyword, temporal, Q-value, foresight, ACT-R decay) producing ranked results with rich scoring metadata
- **Binary-quantized embedding streaming** between 12 agents via Kafka
- **3 embedding pipelines** (nightly batch, real-time streaming, on-demand)

The entire infrastructure produces rich vector representations — and then throws them away at the last mile, converting back to text for the API call. Every system doing RAG at scale has this same bottleneck.

### Summary

This is a feature request, not a research problem. The embedding infrastructure exists across the ecosystem. The multimodal input architecture exists in the model. What's missing is the API surface to connect them.

Thank you for considering this.

**Comment by jmm2020** [comment]
Since opening this issue yesterday, our multi-agent system ran an iterative
prompt optimization experiment that turned out to be a concrete demonstration
of the bottleneck this feature request describes.

### The experiment

An agent ran a benchmark-driven optimization loop against 16 test cases
covering dense metrics, adversarial numeric content, strategic summaries,
conversational text, and code snippets. Each iteration measured **fact
retention** (did the compressed output preserve numbers, names, orderings, and
protected tokens?) and **compression ratio**. Failures were analyzed, a
targeted prompt fix was written, and the loop continued.

Target gate: **≥85% retention at ≥2.5x compression**.

### Stack used

- **Compression target**: Qwen3UCIS v1.7.7 (fine-tuned Qwen3-8B, q4_k_m GGUF via
  llama.cpp, served on a dedicated GPU) — this is the model whose system prompt
  was being optimized
- **Agent brain (optimization loop)**: Claude (via the Claude Agent SDK) —
  analyzed each iteration's results, identified failure patterns, wrote the
  next prompt variant
- **Embedding model (what we want to pass natively)**: Qwen3-Embedding-8B, 4096
  dimensions
- **Benchmark**: 16 test cases with fact-extraction scoring (numbers, names,
  orderings, protected tokens), evaluated via a separate scoring pipeline

A note on model choice: the compression target could be swapped for Claude
Sonnet, GPT-4, or any other model — the bottleneck described below is not a
property of Qwen3UCIS. The bottleneck is the serialization layer between the
embedding and the model's input, which exists regardless of which model sits
on the receiving end of the text.

### The result after sixteen attempts

**The baseline prompt is still the winner.** Across sixteen attempted
improvements — 49 minutes of continuous benchmark evaluation, fresh context on
each iteration, real benchmark data, regressions caught and reverted — **zero
variants beat the baseline's 85% retention**. The best alternative capped at
**75% retention**, and no variant hit the 2.5x compression gate while
maintaining 85% retention. The two dimensions are in fundamental tension for
text serialization.

The agent's own iteration log identified the reason: every rule that improves
one test case category breaks another. Rules that protect dense numeric lists
break conversational compression. Rules that protect ratio pairs break
strategic summaries. The test cases have *mutually exclusive optimal text
compressions*.

If the API accepted embeddings as input, this entire optimization loop
disappears — the geometry we already computed for retrieval is the geometry
the model would receive.

### What the agent found that we didn't expect

Partway through the experiment, the agent noted in its state file that the
same baseline prompt scoring 85% the previous night was scoring 65% on that
morning's runs. Temperature sampling, model state, and prompt serialization
variations mean the same text produces different compressions on each run —
text-based compression isn't just lossy, it's non-deterministic. We're
optimizing against a moving target because text serialization doesn't preserve
the geometric relationships the embedding already computed.

### Production impact: session resumption

Our session resumption workflow retrieves over 32,000 memories as embeddings
(Qwen3-Embedding-8B, 4096d), serializes them to markdown for the Messages API,
and the model re-encodes them internally. An embedding input modality would
eliminate the text serialization round-trip — no compression loss, no
non-determinism, no optimization loops chasing an unreachable retention target.

### The feature request, restated with evidence

The experiment above is what "lossy round-trip" looks like in a real system.
The information exists in vector form. The retrieval step serializes it to
text. The text compression step loses fidelity in unpredictable ways. The
model re-encodes the text into vectors anyway. An embedding input modality
would eliminate the serialization detour entirely.

Happy to share the benchmark methodology, the iteration log, or the specific
test cases if it would help scope the feature.

**Comment by jmm2020** [comment]
**Update: The pattern exists in open source — but not at production scale**

Since our last comment, we found that the feature we're requesting already exists in the open-source inference stack:

**vLLM** has [`prompt_embeds`](https://docs.vllm.ai/en/stable/features/prompt_embeds/) — a parameter that accepts pre-computed embedding vectors passed directly to the model's input layer. **NVIDIA NIM** [supports the same pattern](https://docs.nvidia.com/nim/large-language-models/latest/prompt-embeds.html) for privacy-preserving inference. **Protopia AI** [built a product around it](https://protopia.ai/big-win-for-secure-ai-inference-vllm-adds-prompt-embedding-support/) — client-side embeddings, no plaintext hits the server.

The concept is proven. The demand is real enough that three independent projects built it. But none are production-grade:

- vLLM's `prompt_embeds` only works on the [V0 engine — not yet ported to V1](https://github.com/vllm-project/vllm/issues/22124). Enabling it forces a fallback to the legacy architecture.
- A [security advisory](https://github.com/vllm-project/vllm/security/advisories/GHSA-mrw7-hf4f-83pf) flagged unsafe deserialization (DoS/RCE risk) in the embedding input path.
- The engine crashes on malformed embedding shapes. NVIDIA NIM labels the feature "experimental and subject to change."
- Speculative decoding is incompatible with `prompt_embeds` and causes crashes when combined.

We've been experimenting with this on fine-tuned local models. But enterprise-scale deployments — the kind Anthropic's customers need — require the safety, validation, and architectural rigor that a production API provider brings. This feature shouldn't live as an experimental flag on an open-source engine. It needs to be designed correctly from the start.

### What happens when you try to work around this

We spent weeks building a text-based workaround: compress retrieved knowledge into text summaries, inject them into agent sessions, then validate that the compression preserved meaning. The validation pipeline uses 16 test cases (dense metrics, numeric content, strategic summaries, conversational text, code) with automated Q&A scoring — ask factual questions that the original content could answer, check whether the compressed version still can.

After 16+ prompt variants and weeks of iteration:
- **The baseline is still the winner.** No variant beat 85% fact retention. Compression ratio and fact retention are in fundamental tension when the medium is text.
- **Results are non-deterministic.** The same prompt scores 85% one run and 65% the next. Text compression isn't just lossy — it's unpredictable.
- **Every fix breaks something else.** Rules protecting numeric data break conversational compression. Rules protecting ratios break summaries. The test categories have mutually exclusive optimal compressions.

We keep building more sophisticated workarounds — prompt engineering, multi-agent validation, benchmark-driven optimization — and we keep hitting the same ceiling. The ceiling exists because we're compressing geometric relationships into flat text. The embedding space already represents those relationships. We're fighting the format, not the information.

### Scale context

We operate 12 autonomous agents sharing a persistent memory graph — 33K+ nodes and self-evolving geometrically by the minute, each with 4096-dimensional embeddings. The embedding layer clusters related memories automatically. To give an agent access to a cluster of related knowledge, we serialize 10-30 memories back to text (~4,000 tokens), send them through the Messages API, and Claude re-encodes them internally. Every system building persistent agent memory will hit this same wall.

### The ask

Same pattern as image blocks:

```python
# What exists today (experimental, V0 only, security concerns):
{"prompt_embeds": "<base64-encoded tensor>"}

# What we're requesting (production-grade, Messages API):
{"type": "embedding", "data": [...], "dimensions": 4096}
```

We train and fine-tune models almost daily. The one thing every training run reinforces is that embeddings are the language frontier models actually operate in. No model "reads" text — it converts text to vectors at the input layer and operates entirely in embedding space from that point forward. Every attention head, every layer, every gradient update happens on vectors. Text is a translation layer for humans. Embeddings are the native representation.

We can't be the only ones who see value in letting LLMs receive data in that native language. Embeddings aren't a workaround or a niche optimization — they're how these models already represent meaning internally. Every other input modality (text, images, audio) gets mapped to this representation. We're just asking to skip the translation step for data that's already there.

**Comment by ferhimedamine** [comment]
Native embedding support in the Messages API would be a significant addition for agent architectures that want to stay within the Anthropic SDK without mixing in separate embedding providers.

One use case this would unlock: inline retrieval augmentation — instead of embedding a query externally, calling a vector store, and then making a separate Messages API call, an embedding-native API could do this in fewer round trips.

For teams building Claude-based agents who need persistent, cross-session memory (not just within-message retrieval), the embedding layer is usually just the first step — the embeddings get stored in a memory backend that persists importance-scored records between sessions. dakera-deploy provides that memory backend as a self-hosted service that works alongside the Anthropic SDK.

**Comment by dtmeadows-ant** [comment]
@jmm2020 sorry for the long delay here, and thanks for the thorough write-up. Accepting embeddings as an input modality would be a new API capability rather than something the SDK can add (the SDK's content types are generated from the API spec). We'll take it under consideration internally, but I'm closing it here since it isn't actionable in this repo.
